File: packages/nodejs/whatsapp-sdk/src/graph.ts
import { errorFromStatus } from "./errors.js";
import { asRecord, decodeBody } from "./models.js";
import { DEFAULT_GRAPH_BASE_URL, DEFAULT_GRAPH_VERSION } from "./types.js";
import type { FetchLike, MetaMessageResult, SendMessageInput } from "./types.js";

export interface GraphClientOptions {
  accessToken: string;
  phoneNumberId: string;
  wabaId?: string;
  graphVersion?: string;
  graphBaseUrl?: string;
  fetch?: FetchLike;
  timeoutMs?: number;
}

export class GraphClient {
  private readonly base: string;
  private readonly fetchImpl: FetchLike;
  private readonly timeoutMs: number;

  constructor(private readonly options: GraphClientOptions) {
    const version = options.graphVersion?.trim() || DEFAULT_GRAPH_VERSION;
    const root = (options.graphBaseUrl?.trim() || DEFAULT_GRAPH_BASE_URL).replace(/\/+$/, "");
    this.base = `${root}/${version}/`;
    this.fetchImpl = options.fetch ?? fetch;
    this.timeoutMs = options.timeoutMs && options.timeoutMs > 0 ? options.timeoutMs : 30_000;
  }

  get phoneNumberId(): string {
    return this.options.phoneNumberId;
  }

  get wabaId(): string | undefined {
    return this.options.wabaId;
  }

  async sendMessage(input: SendMessageInput): Promise<MetaMessageResult> {
    const body = buildMessageBody(input);
    const payload = await this.request("POST", `${this.options.phoneNumberId}/messages`, body);
    return parseMessageResult(payload);
  }

  async listTemplates(): Promise<unknown> {
    if (!this.options.wabaId) {
      throw new Error("wabaId e obrigatorio para listar templates");
    }
    return this.request("GET", `${this.options.wabaId}/message_templates`);
  }

  async uploadMedia(bytes: Uint8Array, mimeType: string, filename = "file"): Promise<{ id: string }> {
    const form = new FormData();
    form.set("messaging_product", "whatsapp");
    form.set("type", mimeType);
    form.set("file", new Blob([Buffer.from(bytes)], { type: mimeType }), filename);
    const payload = await this.request(
      "POST",
      `${this.options.phoneNumberId}/media`,
      form,
    );
    const rec = asRecord(payload);
    return { id: String(rec.id ?? "") };
  }

  async getMedia(mediaId: string): Promise<{ url?: string; mimeType?: string; sha256?: string; fileSize?: number; raw: unknown }> {
    const payload = await this.request("GET", mediaId);
    const rec = asRecord(payload);
    return {
      url: rec.url == null ? undefined : String(rec.url),
      mimeType: rec.mime_type == null ? undefined : String(rec.mime_type),
      sha256: rec.sha256 == null ? undefined : String(rec.sha256),
      fileSize: rec.file_size == null ? undefined : Number(rec.file_size),
      raw: payload,
    };
  }

  async request(method: string, path: string, body?: unknown): Promise<unknown> {
    const url = new URL(path.replace(/^\/+/, ""), this.base);
    const headers = new Headers();
    headers.set("Authorization", `Bearer ${this.options.accessToken}`);
    headers.set("Accept", "application/json");

    let encoded: string | FormData | undefined;
    if (body instanceof FormData) {
      encoded = body;
    } else if (body !== undefined && body !== null) {
      encoded = JSON.stringify(body);
      headers.set("Content-Type", "application/json");
    }

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    let response: Response;
    try {
      response = await this.fetchImpl(url, {
        method: method.toUpperCase(),
        headers,
        body: encoded,
        signal: controller.signal,
      });
    } finally {
      clearTimeout(timer);
    }

    const payload = decodeBody(await response.text());
    if (response.ok) return payload;

    const rec = asRecord(payload);
    const err = asRecord(rec.error);
    const message =
      (typeof err.message === "string" && err.message) ||
      (typeof rec.message === "string" && rec.message) ||
      `Erro HTTP ${response.status} na Graph API`;
    throw errorFromStatus(response.status, message, payload, rec.error ?? rec.errors ?? null);
  }
}

export function buildMessageBody(input: SendMessageInput): Record<string, unknown> {
  const body: Record<string, unknown> = {
    messaging_product: "whatsapp",
    recipient_type: "individual",
    to: input.to,
    type: input.type === "template" ? "template" : input.type,
  };

  const callback = buildCallbackData(input);
  if (callback) body.biz_opaque_callback_data = callback;

  if (input.type === "template") {
    if (!input.template?.name) {
      throw new Error("template.name e obrigatorio para envio de template");
    }
    body.template = {
      name: input.template.name,
      language: { code: input.template.language || "pt_BR" },
      components: input.template.components,
    };
    return body;
  }

  if (input.type === "text") {
    if (!input.text?.body) {
      throw new Error("text.body e obrigatorio para mensagem de sessao");
    }
    body.text = {
      preview_url: Boolean(input.text.previewUrl),
      body: input.text.body,
    };
    return body;
  }

  const media = input.media;
  if (!media || (!media.id && !media.link)) {
    throw new Error("media.id ou media.link e obrigatorio");
  }
  const mediaBody: Record<string, unknown> = {};
  if (media.id) mediaBody.id = media.id;
  if (media.link) mediaBody.link = media.link;
  if (media.caption) mediaBody.caption = media.caption;
  if (media.filename) mediaBody.filename = media.filename;
  body[input.type] = mediaBody;
  return body;
}

function buildCallbackData(input: SendMessageInput): string | undefined {
  if (input.callbackData) return input.callbackData.slice(0, 512);
  const payload = JSON.stringify({ category: input.category });
  return payload.length <= 512 ? payload : undefined;
}

export function parseMessageResult(payload: unknown): MetaMessageResult {
  const rec = asRecord(payload);
  const contactsRaw = Array.isArray(rec.contacts) ? rec.contacts : [];
  const messagesRaw = Array.isArray(rec.messages) ? rec.messages : [];
  return {
    messagingProduct: rec.messaging_product == null ? undefined : String(rec.messaging_product),
    contacts: contactsRaw.map((c) => {
      const row = asRecord(c);
      return {
        input: row.input == null ? undefined : String(row.input),
        waId: row.wa_id == null ? undefined : String(row.wa_id),
      };
    }),
    messages: messagesRaw.map((m) => {
      const row = asRecord(m);
      return {
        id: String(row.id ?? ""),
        messageStatus: row.message_status == null ? undefined : String(row.message_status),
      };
    }),
    raw: payload,
  };
}

File: packages/ruby/whatsapp/lib/ordex_whatsapp/graph.rb
# frozen_string_literal: true

require "faraday"
require "json"

module OrdexWhatsApp
  class GraphClient
    def initialize(access_token:, phone_number_id:, waba_id: nil, graph_version: nil, graph_base_url: nil, connection: nil)
      version = graph_version.to_s.strip.empty? ? DEFAULT_GRAPH_VERSION : graph_version
      root = (graph_base_url.to_s.strip.empty? ? DEFAULT_GRAPH_BASE_URL : graph_base_url).sub(%r{/$}, "")
      @access_token = access_token
      @phone_number_id = phone_number_id
      @waba_id = waba_id
      @connection = connection || Faraday.new(url: "#{root}/#{version}/") do |f|
        f.request :json
        f.adapter Faraday.default_adapter
      end
    end

    def send_message(input)
      body = self.class.build_message_body(input)
      payload = request(:post, "#{@phone_number_id}/messages", body)
      self.class.parse_message_result(payload)
    end

    def self.build_message_body(input)
      input = stringify(input)
      body = {
        "messaging_product" => "whatsapp",
        "recipient_type" => "individual",
        "to" => input["to"],
        "type" => input["type"]
      }
      callback = callback_data(input)
      body["biz_opaque_callback_data"] = callback if callback

      if input["type"] == "template"
        template = stringify(input["template"])
        raise ConfigurationError, "template.name e obrigatorio para envio de template" if template["name"].to_s.strip.empty?

        body["template"] = {
          "name" => template["name"],
          "language" => { "code" => template["language"].to_s.empty? ? "pt_BR" : template["language"] },
          "components" => template["components"]
        }.compact
        return body
      end

      if input["type"] == "text"
        text = stringify(input["text"])
        raise ConfigurationError, "text.body e obrigatorio para mensagem de sessao" if text["body"].to_s.strip.empty?

        body["text"] = { "preview_url" => !!text["preview_url"], "body" => text["body"] }
        return body
      end

      media = stringify(input["media"])
      raise ConfigurationError, "media.id ou media.link e obrigatorio" if media["id"].to_s.empty? && media["link"].to_s.empty?

      media_body = {}
      media_body["id"] = media["id"] if media["id"]
      media_body["link"] = media["link"] if media["link"]
      media_body["caption"] = media["caption"] if media["caption"]
      media_body["filename"] = media["filename"] if media["filename"]
      body[input["type"]] = media_body
      body
    end

    def self.parse_message_result(payload)
      rec = payload.is_a?(Hash) ? payload : {}
      {
        messaging_product: rec["messaging_product"],
        contacts: Array(rec["contacts"]).map { |c| { input: c["input"], wa_id: c["wa_id"] } },
        messages: Array(rec["messages"]).map { |m| { id: m["id"].to_s, message_status: m["message_status"] } },
        raw: payload
      }
    end

    def self.callback_data(input)
      return input["callback_data"].to_s[0, 512] if input["callback_data"]

      payload = JSON.dump({ "category" => input["category"] })
      payload.length <= 512 ? payload : nil
    end

    def self.stringify(hash)
      return {} if hash.nil?
      return hash.transform_keys(&:to_s) if hash.is_a?(Hash)

      {}
    end

    private

    def request(method, path, body = nil)
      resposta = @connection.run_request(method, path, nil, {}) do |req|
        req.headers["Authorization"] = "Bearer #{@access_token}"
        req.headers["Accept"] = "application/json"
        if body
          req.headers["Content-Type"] = "application/json"
          req.body = JSON.dump(body)
        end
      end
      payload = decode(resposta.body)
      return payload if resposta.success?

      rec = payload.is_a?(Hash) ? payload : {}
      err = rec["error"].is_a?(Hash) ? rec["error"] : {}
      mensagem = err["message"] || rec["message"] || "Erro HTTP #{resposta.status} na Graph API"
      raise OrdexWhatsApp.error_from_status(resposta.status, mensagem, body: payload, errors: rec["error"])
    end

    def decode(corpo)
      return {} if corpo.nil? || corpo.to_s.strip.empty?

      JSON.parse(corpo)
    rescue JSON::ParserError
      { "raw" => corpo }
    end
  end
end
